lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1103,6 +1103,29 @@ emitSIBIFNeededAfterOptimizations(CompilerInvocation &Invocation, SILModule *SM,
              : serializeMSF(opts, SM, Context, MSF);
 }
 
+static Optional<bool> emitModuleIfNeeded(SILModule *SM,
+                                         CompilerInvocation &Invocation,
+                                         CompilerInstance &Instance,
+                                         ModuleOrSourceFile MSF) {
+  const FrontendOptions &opts = Invocation.getFrontendOptions();
+  if (opts.ModuleOutputPath.empty() && opts.ModuleDocOutputPath.empty())
+    return None;
+
+  // Serialize the SILModule if it was not serialized yet.
+  if (!SM->isSerialized())
+    SM->serialize();
+  auto action = Invocation.getFrontendOptions().RequestedAction;
+  if (action != FrontendOptions::ActionType::MergeModules &&
+      action != FrontendOptions::ActionType::EmitModuleOnly)
+    return None;
+
+  if (Invocation.getFrontendOptions().IndexStorePath.empty())
+    return Instance.getASTContext().hadError();
+
+  return emitIndexData(MSF.dyn_cast<SourceFile *>(), Invocation, Instance) ||
```

Comment:
This doesn't exactly fit with the name 'emitModuleIfNeeded'. It's true that index data isn't *useful* without a module, but…

@huonw, comments?

Reply:
Refactored differently, into ensureSILModuleIsSerialized and adding IfNeeded to emitIndexData.